### PR TITLE
[Enhancement] Reduce cpu cost in pk publish

### PR DESCRIPTION
**Pull request by TszKitLo40**
## Why I'm doing:
The constructor `explicit Rowset(TabletManager* tablet_mgr, TabletMetadataPtr tablet_metadata, int rowset_index);` will emplace schema to `GlobalTabletSchemaMap`. But the emplace of `GlobalTabletSchemaMap` need lock which will waste cpu cost. And in publish version, PK table needs to pick rowsets first and calculate compaction score, but only rowset indexes is needed in fact which means there is no need to construct `Rowset`.
## What I'm doing:
Split the logic of `pick_rowset_indexes` from `pick_rowsets`. For compaction score, we only need rowset indexes. For compaction, rowset can be constructed by rowset indexes.
Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/storage/lake/compaction_policy.cpp [hunk 1]
```diff
@@ -134,6 +134,8 @@ class PrimaryCompactionPolicy : public CompactionPolicy {
     StatusOr<std::vector<RowsetPtr>> pick_rowsets() override;
     StatusOr<std::vector<RowsetPtr>> pick_rowsets(const std::shared_ptr<const TabletMetadataPB>& tablet_metadata,
                                                   bool calc_score, std::vector<bool>* has_dels);
+    StatusOr<std::vector<int64_t>> pick_rowset_indexes(const std::shared_ptr<const TabletMetadataPB>& tablet_metadata,
```
Comment (luohaha) [inline be/src/storage/lake/compaction_policy.cpp:137]: Make it private

Comment (luohaha) [inline be/src/storage/lake/compaction_policy.cpp:137]: Add some comments about what does this function return, and why we need this.

Comment (TszKitLo40) [reply be/src/storage/lake/compaction_policy.cpp:140]: done

**Review by luohaha — DISMISSED** [review]

File: be/src/storage/lake/compaction_policy.cpp [hunk 2]
```diff
@@ -171,32 +175,40 @@ StatusOr<std::vector<RowsetPtr>> PrimaryCompactionPolicy::pick_rowsets(
...
+}
+
+StatusOr<std::vector<RowsetPtr>> PrimaryCompactionPolicy::pick_rowsets(
+        const std::shared_ptr<const TabletMetadataPB>& tablet_metadata, bool calc_score, std::vector<bool>* has_dels) {
+    std::vector<RowsetPtr> input_rowsets;
+    ASSIGN_OR_RETURN(auto rowset_indexes, pick_rowset_indexes(tablet_metadata, calc_score, has_dels));
+    std::stringstream input_infos;
+    for (auto rowset_index : rowset_indexes) {
+        input_rowsets.emplace_back(std::make_shared<Rowset>(_tablet_mgr, tablet_metadata, rowset_index));
+        input_infos << tablet_metadata->rowsets(rowset_index).id() << "|";
```
Comment (sduzh) [inline be/src/storage/lake/compaction_policy.cpp:208]: Use the `JoinMapped` function.

File: be/src/storage/lake/compaction_policy.cpp [hunk 3]
```diff
@@ -171,32 +175,42 @@ StatusOr<std::vector<RowsetPtr>> PrimaryCompactionPolicy::pick_rowsets(
...
+}
+
+StatusOr<std::vector<RowsetPtr>> PrimaryCompactionPolicy::pick_rowsets(
+        const std::shared_ptr<const TabletMetadataPB>& tablet_metadata, bool calc_score, std::vector<bool>* has_dels) {
+    std::vector<RowsetPtr> input_rowsets;
+    ASSIGN_OR_RETURN(auto rowset_indexes, pick_rowset_indexes(tablet_metadata, calc_score, has_dels));
+    std::string input_infos = JoinMapped(
+            rowset_indexes,
+            [&](int64_t rowset_index) -> std::string {
+                input_rowsets.emplace_back(std::make_shared<Rowset>(_tablet_mgr, tablet_metadata, rowset_index));
```
Comment (sduzh) [inline be/src/storage/lake/compaction_policy.cpp:208]: Do not fill input_rowsts in `JoinMapper`

File: be/src/storage/lake/compaction_policy.cpp [hunk 4]
```diff
@@ -171,32 +175,42 @@ StatusOr<std::vector<RowsetPtr>> PrimaryCompactionPolicy::pick_rowsets(
...
+                return std::to_string(tablet_metadata->rowsets(rowset_index).id());
+            },
+            "|");
     VLOG(2) << strings::Substitute("lake PrimaryCompactionPolicy pick_rowsets tabletid:$0 version:$1 inputs:$2",
-                                   tablet_id, tablet_metadata->version(), input_infos.str());
+                                   tablet_metadata->id(), tablet_metadata->version(), input_infos);
```
Comment (sduzh) [inline be/src/storage/lake/compaction_policy.cpp:213]:
```suggestion
                                   tablet_metadata->id(), tablet_metadata->version(), JoinMapped(....));
```

Comment (TszKitLo40) [reply be/src/storage/lake/compaction_policy.cpp]: done

File: be/src/storage/lake/compaction_policy.cpp [hunk 5]
```diff
@@ -171,46 +175,57 @@ StatusOr<std::vector<RowsetPtr>> PrimaryCompactionPolicy::pick_rowsets(
...
-    VLOG(2) << strings::Substitute("lake PrimaryCompactionPolicy pick_rowsets tabletid:$0 version:$1 inputs:$2",
-                                   tablet_id, tablet_metadata->version(), input_infos.str());
+    return rowset_indexes;
+}
 
+StatusOr<std::vector<RowsetPtr>> PrimaryCompactionPolicy::pick_rowsets(
+        const std::shared_ptr<const TabletMetadataPB>& tablet_metadata, bool calc_score, std::vector<bool>* has_dels) {
+    std::vector<RowsetPtr> input_rowsets;
+    ASSIGN_OR_RETURN(auto rowset_indexes, pick_rowset_indexes(tablet_metadata, calc_score, has_dels));
+    for (auto rowset_index : rowset_indexes) {
```
Comment (sduzh) [inline be/src/storage/lake/compaction_policy.cpp:205]:
```suggestion
    input_rowsets.reserve(rowset_indexes.size());
    for (auto rowset_index : rowset_indexes) {
```

Comment (TszKitLo40) [reply be/src/storage/lake/compaction_policy.cpp:206]: done

File: be/src/storage/lake/compaction_policy.cpp [hunk 6]
```diff
@@ -171,46 +175,58 @@ StatusOr<std::vector<RowsetPtr>> PrimaryCompactionPolicy::pick_rowsets(
...
+        input_rowsets.emplace_back(std::make_shared<Rowset>(_tablet_mgr, tablet_metadata, rowset_index));
+    }
+    VLOG(2) << strings::Substitute("lake PrimaryCompactionPolicy pick_rowsets tabletid:$0 version:$1 inputs:$2",
+                                   tablet_metadata->id(), tablet_metadata->version(),
+                                   JoinMapped(
+                                           rowset_indexes,
+                                           [&](int64_t rowset_index) -> std::string {
+                                               return std::to_string(tablet_metadata->rowsets(rowset_index).id());
+                                           },
+                                           "|"));
```
Comment (sduzh) [inline be/src/storage/lake/compaction_policy.cpp:216]:
```suggestion
                                   JoinMapped(
                                           input_rowsets,
                                           [&](const RowsetPtr& rowset) -> std::string {
                                               return std::to_string(rowset->id());
                                           },
                                           "|"));
```

**Review by sduzh — APPROVED** [review]

**Comment by sduzh** [comment]
LGTM, PTAL @luohaha

**Review by luohaha — APPROVED** [review]

**Pull request merged**